content/articles/building-a-custom-stopwatch-npm-library-in-typescript/index.md

```diff
@@ -0,0 +1,355 @@
+
+In the `src` folder we will create the `stopwatch.ts` file. Once that is done, let’s dive right into the stopwatch setup.
+
+In the `stopwatch.ts` file, implement the code snippet below:
+
+```typescript
+export function startWatch(timerStop: boolean, watchCycle: any) {
+  if (timerStop) {
+    timerStop = false;
+    watchCycle();
+  }
+}
+
+export function stopWatch(timerStop: boolean) {
+  if (timerStop == false) {
+    timerStop = true;
+  }
+}
+
+export function watchCycle(
+  timerStop: boolean,
+  milliseconds: number,
+  seconds: number,
+  minutes: number,
+  hours: number,
+  timer: HTMLElement
+): number | string {
+  if (timerStop == false) {
+    milliseconds = milliseconds + 10;
+
+    if (milliseconds == 1000) {
+      milliseconds = 0;
+      seconds = seconds + 1;
+    }
+    if (seconds == 60) {
+      minutes = minutes + 1;
+      seconds = 0;
+    }
+    if (minutes == 60) {
+      hours = hours + 1;
+      minutes = 0;
+      seconds = 0;
+    }
+
+    let hr = hours < 10 ? "0" + hours : hours;
+    let mins = minutes < 10 ? "0" + minutes : minutes;
+    let secs = seconds < 10 ? "0" + seconds : seconds;
+    let milsec =
+      milliseconds < 10
+        ? "00" + milliseconds
+        : milliseconds < 100
+        ? "0" + milliseconds
+        : milliseconds;
+
+    setTimeout("watchCycle()", 10);
+    return (timer.innerHTML = hr + ":" + mins + ":" + secs + ":" + milsec);
+  }
+}
+
+export function resetWatch(
+  milliseconds: number,
+  seconds: number,
+  minutes: number,
+  hours: number,
+  timer: HTMLElement
+) {
+  hours = 0;
+  minutes = 0;
+  seconds = 0;
+  milliseconds = 0;
+  return (timer.innerHTML = "00:00:00");
+}
+```
+
+As you are already familiar with Typescript, we are expressly assigning a value type to every parameter each function is accepting (i.e. number, string, HTMLElement, and any).
+
+From the snippet above, we created and exported 4 functions as explained below:
+
+- `startWatch`: This function is called when the `start` button is clicked. On execution, the function sets the `timerStop` to false and also starts the `watchCycle` function.
+- `stopWatch`: This function runs when the user clicks the `stop` button. When executed, the `stopwatch` function sets the `timerstop` to true which halts the `watchCycle` function.
+- `resetWatch`: The `resetWatch` function simply returns the stopwatch with the initial value (0) and it is executed when the user clicks the `reset` button.
+- `watchCycle`: This is the brain behind our stopwatch, it increments the values of the milliseconds, seconds, minutes, and hours accordingly and then returns the values to the `timer` HTML element tag in the `index.html` when used.
+
+That is all we need to set up the stopwatch logic. Finally, create an `index.ts` in `src` and export all the functions as shown below:
+
+```typescript
+export { startWatch, watchCycle, stopWatch, resetWatch } from "./stopwatch";
+```
+
+### Step 6 – Testing/consuming the stopwatch locally

```

Comment:
@davekola 

In this step, you have re-iplemented your stopwatch library. This doen't show the reader how to use the typescript library. Please don't re-implement the library. Instead, you can transpile the typescript to javascript and use it using the import statement.

 `import StopWatch from ...`